File: code-contests-dataset/solutions/qwen3-cpp/coci11c1p3.cpp
#include <iostream>
#include <parlay/sequence.h>
#include <parlay/primitives.h>

using namespace std;

int main() {
    int N;
    cin >> N;
    
    parlay::sequence<long long> numbers(N);
    for (int i = 0; i < N; i++) {
        cin >> numbers[i];
    }
    
    long long result = 0;
    
    // For each bit position, count how many numbers have 0 and how many have 1
    // The contribution to the final sum is the product of these counts
    for (int bit = 0; bit < 30; bit++) { // 30 bits is enough for numbers < 1000000
        // Create a sequence indicating whether each number has bit set to 1
        auto bit_flags = parlay::map(numbers, [bit](long long x) {
            return (x >> bit) & 1;
        });
        
        // Count zeros and ones in parallel
        long long count_1 = parlay::reduce(bit_flags, parlay::plus<long long>());
        long long count_0 = N - count_1;
        
        // For this bit position, the total contribution is count_0 * count_1
        result += count_0 * count_1 * (1LL << bit);
    }
    
    cout << result << endl;
    return 0;
}

File: code-contests-dataset/solutions/qwen3-rust/coci11c1p3.rs
use std::io::{self, Read};
use rayon::prelude::*;

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    
    let mut lines = input.lines();
    let n: usize = lines.next().unwrap().parse().unwrap();
    let mut numbers: Vec<u32> = Vec::with_capacity(n);
    
    for _ in 0..n {
        numbers.push(lines.next().unwrap().parse().unwrap());
    }
    
    // For each bit position, count how many numbers have 0 and 1
    let mut bit_counts = vec![(0u64, 0u64); 20]; // 20 bits enough for numbers < 1_000_000
    
    for &num in &numbers {
        for i in 0..20 {
            if num & (1 << i) != 0 {
                bit_counts[i].1 += 1;
            } else {
                bit_counts[i].0 += 1;
            }
        }
    }
    
    // Calculate contribution of each bit position to total sum
    let result: u64 = bit_counts
        .par_iter()
        .enumerate()
        .map(|(i, &(zeros, ones))| {
            // For each bit position, the friendship value will be 1 if bits differ
            // Number of pairs with differing bits = zeros * ones
            zeros * ones * (1u64 << i)
        })
        .sum();
    
    println!("{}", result);
}